### PR TITLE
cleanup code, add error handling, introduce packer & add a varint implementation and break a lot of stuff :D

### PR DESCRIPTION
**Pull request by jxsl13**
This is a draft
I broke many tests :D
I just wanted to visualize some potential improvements where you can deduplicate code a little bit and where you could add more error handling.

I would suggest not to use value receivers `func (b Blub) MethodName()`, as they are rarely needed, primarily with smaller primitive types, as they cause your whole object to be copied whereas pointer receivers only copy a pointer (on the stack).

In order to keep your objects compact (and locally together), another suggestion of mine would be to make all ChunkHeaders actual value types instead of pointers.

A lot of the errors that are simply passed back to the caller could be improved a lot in this draft, in order to actually see the error trace when for example unpacking of a chunk fails.

Every slice access should be guarded, in order to avoid index out of range panics, see where the chunk header is parsed.

In order to improve performance, it might at some point in the future make sense to check out how Go's `io.Reader` and `io.Writer` are implemented in order to avoid too many heap allocations with `make([]byte, 0, 0)`. This is especially interesting for all `Pack()` methods.

### CODE REVIEW

File: messages7/token.go [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package messages7
+
+var TokenPadding = []byte{512: 0}
```
Comment (ChillerDragon) [inline messages7/token.go:3]: That should probably go to ctrl_token.go

**Review by ChillerDragon — CHANGES_REQUESTED** [review]

**Comment by ChillerDragon** [comment]
Amazing thanks for this ❤️❤️❤️

Its sadly a huge diff with lots of different things so I can not merge it as is since i disagree with a few choices.

Things I would love to merge straight away are:
- The value receivers being changed to pointer receivers. They were value receivers because I had go skill issues getting the interface to work.
- The the msg cast switch statement that removed my mad duplicated code.

Stuff I am not so sure about:
- Adding external dependencies. The current test code is horribly verbose and there has to be a cleaner way. But at least it is using the go standard library. I understand that go has a working dependency system and using other peoples code is encouraged. But I personally like to try avoiding dependencies. Especially while exploring a new language. The test code could also be cleaned up with a simple self rolled assert method. No need to add an entire testing library dependency that supports mocking and brings a bunch of peer dependencies.
- I actually made the ChunkHeader intentionally a pointer. To have an optional value that can be nil. To allow auto generating chunk headers and manually setting them. So a user of the library can do those two things:
```go
// header is nil so it will be auto generated in Pack() setting the correct size
packet := Packet{}
packet.Messages.append = append(packet.Messages, &messages7.SvChat{Message: "foo"})
packet.Pack()
```

```go
// header is manually set by the user so the size will not be recomputed even if it is wrong
packet := Packet{}
packet.Messages.append = append(
  packet.Messages,
  &messages7.SvChat{ChunkHeader: &ChunkHeader {Size: 9999999}, Message: "foo"},
)
packet.Pack()
```
- Error handling is nice it was on my todo. The word Next makes sense but I would prefer keeping GetInt() instead of changing it to NextInt() just because I am used to that already from the C++ code.
- The new packer initialization seems complex. I assume it is probably faster than mine because of the batched memory allocation. But the code becomes so much more verbose. Harder to read, harder to write, harder to debug. This is something I would like to delay until the project is more matured. If a more complex code base makes me rage quit the project then faster runtime helps nobody. I have to admit I never tried to learn go this is my first project using it and I did little to no research on how to write good/fast/idiomatic go. I chose the first solution that works/compiles/looks okay after a bit try and error. So I am sure there is lots of room for memory management improvements. But they should as of right now not sacrafice in code complexity. Because I am not even sure if speed is in scope. I personally might only use this project for chat bots or toy projects. But I will use it for nothing if the code is more complex than the C++ reference implementation. If the library becomes more polished and feature complete and someone has an actual use case that depends on speed where writing complex code has noticable effects then we can revisit this one.

**Comment by jxsl13** [comment]
The packer inizialization is optimized for memory allocation.
You could simply do a NewPacker() and work with the returned variable.

It is just me passing an approximated buffer that the buffer is supposed to use.

**Comment by ChillerDragon** [comment]
Ah okay you convinced me. So nothing breaks if the length is computed wrong and it also does not have to be provided at all. So adding new messages is still as smooth as before. That seems nice. Looked more scary at first.

**Comment by ChillerDragon** [comment]
A benchmark would still be interesting. I just assumed that slices.Concat is highly optimized already and that self rolled solutions probably end up with similiar speeds if not worse. I mean slices.Concat could do the same. It probably also only does one memory allocation if you give it 5 slices.

**Comment by ChillerDragon** [comment]
I am having a hard time properly checking the full diff here. Now that https://github.com/teeworlds-go/go-teeworlds-protocol/pull/3 is merged is there still something that should get in in your opinion?

**Comment by jxsl13** [comment]
@ChillerDragon  I'd love for varint to be added:
https://github.com/jxsl13/twapi/blob/master/compression/varint.go
plus tests, which is why I introduced more test utilities:
The introduction of Varint also introduces way more required error handling.

fun fact, the varint api is oriented on the varint implementation in the standard library (which differs from the teeworlds one: https://pkg.go.dev/encoding/binary#AppendVarint

**Comment by ChillerDragon** [comment]
Wouldn't the go way be to just add your code as a dependency instead of copy pasting it? Since individual modules can be loaded from other repos it seems weird to copy paste over your code. I don't like external dependencies with sub dependencies of some third party projects I don't fully understand. But teeworlds specific dependencies are nice. I mean before we literally copy paste the entire file might as well just link it :D

I am not sure if i can get used to the naming with varint and stuff. Also your tests depend on external dependencies so if we would add yours as a depenency here it would bring the bloated test lib as a sub dependency.

I will keep it in my by leaving the pr open. I was planning to also add more error handling and once I get to it I will revisit this idea. But can't make any promises. Currently I lean towards keeping it as is and add errors manually.

**Pull request closed without merging**